weather/crud.py

```diff
@@ -0,0 +1,108 @@
+from fastapi import HTTPException
+
+from sqlalchemy.orm import Session
+from sqlalchemy.future import select
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.exc import IntegrityError
+
+from weather import models, schemas
+from weather.utils import exception_handler
+
+
+async def get_all_cities(db: AsyncSession):
+    result = await db.execute(
+        select(models.City)
+    )
+    return result.scalars().all()
+
+
+@exception_handler
+async def get_city_by_id(db: AsyncSession, city_id: int):
+    result = await db.execute(
+        select(models.City).filter(models.City.id == city_id)
+    )
+    city = result.scalars().first()
+    if not city:
+        raise HTTPException(status_code=404, detail="City not found")
+
+    return city
+
+
+@exception_handler
+async def get_city_by_name(db: AsyncSession, city_name: str):
+    result = await db.execute(
+        select(models.City).filter(models.City.name == city_name)
+    )
+    city = result.scalars().first()
+
+    return city
+
+
+@exception_handler
+async def create_city(db: AsyncSession, city: schemas.CityCreate):
+    if await get_city_by_name(db=db, city_name=city.name):
+        raise HTTPException(status_code=400, detail="City already exists")
```

Comment:
The HTTP status code for an existing city is set to 400. This is appropriate for indicating a bad request due to a conflict with existing data.